File: android/app/src/main/java/com/parksence/api/ApiClient.kt
package com.parksense.api

import android.graphics.Bitmap
import android.util.Base64
import org.json.JSONArray
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

data class ParkingResult(
    val canPark: Boolean?,
    val message: String,
    val notes: List<String>,
    val signs: List<String>,
)

data class UserProfile(
    val id: Int,
    val email: String,
    val vehicleType: String,
    val isDisabled: Boolean,
    val hasResidentPermit: Boolean,
    val residentZone: String,
)

object ApiClient {

    var serverUrl = "http://192.168.68.101:8000"
    var authToken: String? = null

    // ── Auth calls ────────────────────────────────────────────────────────────

    fun login(email: String, password: String): Pair<String, UserProfile> {
        val body = JSONObject().apply {
            put("email", email)
            put("password", password)
        }
        val response = post("/api/auth/login", body.toString(), contentType = "application/json")
        return parseAuthResponse(response)
    }

    fun register(
        email: String,
        password: String,
        vehicleType: String,
        isDisabled: Boolean,
        hasResidentPermit: Boolean,
        residentZone: String,
    ): Pair<String, UserProfile> {
        val body = JSONObject().apply {
            put("email", email)
            put("password", password)
            put("vehicle_type", vehicleType)
            put("is_disabled", isDisabled)
            put("has_resident_permit", hasResidentPermit)
            put("resident_zone", residentZone)
        }
        val response = post("/api/auth/register", body.toString(), contentType = "application/json", expectedCode = 201)
        return parseAuthResponse(response)
    }

    fun updateProfile(
        vehicleType: String,
        isDisabled: Boolean,
        hasResidentPermit: Boolean,
        residentZone: String,
    ): UserProfile {
        val body = JSONObject().apply {
            put("vehicle_type", vehicleType)
            put("is_disabled", isDisabled)
            put("has_resident_permit", hasResidentPermit)
            put("resident_zone", residentZone)
        }
        val response = put("/api/auth/me", body.toString())
        return parseUserProfile(JSONObject(response))
    }

    // ── Analysis ──────────────────────────────────────────────────────────────

    fun analyze(bitmap: Bitmap, dayName: String, timeStr: String): ParkingResult {
        val jpeg = bitmapToJpegBytes(bitmap)

        val boundary = "----ParkSenseBoundary"
        val baos = ByteArrayOutputStream()
        fun write(s: String) = baos.write(s.toByteArray())

        // image field
        write("--$boundary\r\n")
        write("Content-Disposition: form-data; name=\"image\"; filename=\"sign.jpg\"\r\n")
        write("Content-Type: image/jpeg\r\n\r\n")
        baos.write(jpeg)
        write("\r\n")

        // day field
        write("--$boundary\r\n")
        write("Content-Disposition: form-data; name=\"day\"\r\n\r\n")
        write(dayName)
        write("\r\n")

        // time field
        write("--$boundary\r\n")
        write("Content-Disposition: form-data; name=\"time\"\r\n\r\n")
        write(timeStr)
        write("\r\n")

        write("--$boundary--\r\n")

        val response = postMultipart("/api/analyze", baos.toByteArray(), boundary)
        return parseResult(response)
    }

    // ── HTTP helpers ──────────────────────────────────────────────────────────

    private fun post(
        path: String,
        body: String,
        contentType: String = "application/json",
        expectedCode: Int = 200,
    ): String {
        val conn = (URL("$serverUrl$path").openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            setRequestProperty("Content-Type", contentType)
            authToken?.let { setRequestProperty("Authorization", "Bearer $it") }
            doOutput = true
            connectTimeout = 10_000
            readTimeout = 30_000
        }
        OutputStreamWriter(conn.outputStream).use { it.write(body) }
        return readResponse(conn, expectedCode)
    }

    private fun put(path: String, body: String): String {
        val conn = (URL("$serverUrl$path").openConnection() as HttpURLConnection).apply {
            requestMethod = "PUT"
            setRequestProperty("Content-Type", "application/json")
            authToken?.let { setRequestProperty("Authorization", "Bearer $it") }
            doOutput = true
            connectTimeout = 10_000
            readTimeout = 30_000
        }
        OutputStreamWriter(conn.outputStream).use { it.write(body) }
        return readResponse(conn, 200)
    }

    private fun postMultipart(path: String, body: ByteArray, boundary: String): String {
        val conn = (URL("$serverUrl$path").openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            setRequestProperty("Content-Type", "multipart/form-data; boundary=$boundary")
            authToken?.let { setRequestProperty("Authorization", "Bearer $it") }
            doOutput = true
            connectTimeout = 15_000
            readTimeout = 90_000
        }
        conn.outputStream.use { it.write(body) }
        return readResponse(conn, 200)
    }

    private fun readResponse(conn: HttpURLConnection, expectedCode: Int): String {
        val code = conn.responseCode
        val stream = if (code in 200..299) conn.inputStream else conn.errorStream
        val text = stream?.bufferedReader()?.readText() ?: ""
        conn.disconnect()
        if (code != expectedCode) {
            val detail = runCatching { JSONObject(text).optString("detail", text) }.getOrDefault(text)
            throw Exception(detail)
        }
        return text
    }

    // ── Parsers ───────────────────────────────────────────────────────────────

    private fun parseAuthResponse(raw: String): Pair<String, UserProfile> {
        val json = JSONObject(raw)
        val token = json.getString("access_token")
        val profile = parseUserProfile(json.getJSONObject("user"))
        return token to profile
    }

    private fun parseUserProfile(json: JSONObject) = UserProfile(
        id = json.getInt("id"),
        email = json.getString("email"),
        vehicleType = json.optString("vehicle_type", "car"),
        isDisabled = json.optBoolean("is_disabled", false),
        hasResidentPermit = json.optBoolean("has_resident_permit", false),
        residentZone = json.optString("resident_zone", ""),
    )

    private fun parseResult(raw: String): ParkingResult {
        val json = JSONObject(raw)
        val canPark = if (json.isNull("can_park")) null else json.getBoolean("can_park")

        fun jsonArrayToList(key: String) = buildList {
            json.optJSONArray(key)?.let { arr ->
                for (i in 0 until arr.length()) {
                    val item = arr.get(i)
                    when (item) {
                        is JSONObject -> {
                            val text = item.optString("text", "")
                            val desc = item.optString("description", "")
                            if (text.isNotEmpty() && desc.isNotEmpty()) add("$text - $desc")
                            else if (desc.isNotEmpty()) add(desc)
                            else if (text.isNotEmpty()) add(text)
                        }
                        is String -> add(item)
                        else -> add(item.toString())
                    }
                }
            }
        }

        return ParkingResult(
            canPark = canPark,
            message = json.optString("message", ""),
            notes = jsonArrayToList("notes"),
            signs = jsonArrayToList("signs"),
        )
    }

    private fun bitmapToJpegBytes(bitmap: Bitmap): ByteArray {
        val max = 1024
        val scale = minOf(1f, max.toFloat() / maxOf(bitmap.width, bitmap.height))
        val b = if (scale < 1f)
            Bitmap.createScaledBitmap(bitmap, (bitmap.width * scale).toInt(), (bitmap.height * scale).toInt(), true)
        else bitmap
        val out = ByteArrayOutputStream()
        b.compress(Bitmap.CompressFormat.JPEG, 85, out)
        if (b != bitmap) b.recycle()
        return out.toByteArray()
    }
}

File: ios/ParkSence/ParkSence/API/ApiClient.swift
import Foundation
import UIKit

// MARK: - Models

struct ParkingResult {
    let canPark: Bool?
    let message: String
    let notes: [String]
    let signs: [String]
}

struct UserProfile: Codable {
    let id: Int
    let email: String
    let vehicleType: String
    let isDisabled: Bool
    let hasResidentPermit: Bool
    let residentZone: String

    enum CodingKeys: String, CodingKey {
        case id, email
        case vehicleType       = "vehicle_type"
        case isDisabled        = "is_disabled"
        case hasResidentPermit = "has_resident_permit"
        case residentZone      = "resident_zone"
    }
}

// MARK: - Client

final class ApiClient {
    static let shared = ApiClient()
    private init() {}

    var serverUrl = "http://192.168.68.101:8000"
    var authToken: String?

    // MARK: Auth

    func login(email: String, password: String) async throws -> (String, UserProfile) {
        let body: [String: Any] = ["email": email, "password": password]
        let data = try await post(path: "/api/auth/login", json: body)
        return try parseAuthResponse(data)
    }

    func register(
        email: String,
        password: String,
        vehicleType: String,
        isDisabled: Bool,
        hasResidentPermit: Bool,
        residentZone: String
    ) async throws -> (String, UserProfile) {
        let body: [String: Any] = [
            "email": email,
            "password": password,
            "vehicle_type": vehicleType,
            "is_disabled": isDisabled,
            "has_resident_permit": hasResidentPermit,
            "resident_zone": residentZone,
        ]
        let data = try await post(path: "/api/auth/register", json: body, expectedCode: 201)
        return try parseAuthResponse(data)
    }

    func updateProfile(
        vehicleType: String,
        isDisabled: Bool,
        hasResidentPermit: Bool,
        residentZone: String
    ) async throws -> UserProfile {
        let body: [String: Any] = [
            "vehicle_type": vehicleType,
            "is_disabled": isDisabled,
            "has_resident_permit": hasResidentPermit,
            "resident_zone": residentZone,
        ]
        let data = try await put(path: "/api/auth/me", json: body)
        let json = try JSONSerialization.jsonObject(with: data) as! [String: Any]
        return try parseUserProfile(json)
    }

    // MARK: Analysis

    func analyze(image: UIImage, dayName: String, timeStr: String) async throws -> ParkingResult {
        let jpeg = resizedJpeg(image)
        let boundary = "----ParkSenseBoundary"
        var body = Data()

        func append(_ string: String) {
            body.append(Data(string.utf8))
        }

        // image field
        append("--\(boundary)\r\n")
        append("Content-Disposition: form-data; name=\"image\"; filename=\"sign.jpg\"\r\n")
        append("Content-Type: image/jpeg\r\n\r\n")
        body.append(jpeg)
        append("\r\n")

        // day field
        append("--\(boundary)\r\n")
        append("Content-Disposition: form-data; name=\"day\"\r\n\r\n")
        append(dayName)
        append("\r\n")

        // time field
        append("--\(boundary)\r\n")
        append("Content-Disposition: form-data; name=\"time\"\r\n\r\n")
        append(timeStr)
        append("\r\n")

        append("--\(boundary)--\r\n")

        let data = try await request(
            path: "/api/analyze",
            method: "POST",
            body: body,
            contentType: "multipart/form-data; boundary=\(boundary)",
            expectedCode: 200,
            timeout: 90
        )
        return try parseResult(data)
    }

    // MARK: HTTP helpers

    private func post(path: String, json: [String: Any], expectedCode: Int = 200) async throws -> Data {
        let body = try JSONSerialization.data(withJSONObject: json)
        return try await request(
            path: path,
            method: "POST",
            body: body,
            contentType: "application/json",
            expectedCode: expectedCode
        )
    }

    private func put(path: String, json: [String: Any]) async throws -> Data {
        let body = try JSONSerialization.data(withJSONObject: json)
        return try await request(
            path: path,
            method: "PUT",
            body: body,
            contentType: "application/json",
            expectedCode: 200
        )
    }

    private func request(
        path: String,
        method: String,
        body: Data,
        contentType: String,
        expectedCode: Int,
        timeout: TimeInterval = 30
    ) async throws -> Data {
        guard let url = URL(string: "\(serverUrl)\(path)") else {
            throw URLError(.badURL)
        }
        var req = URLRequest(url: url, timeoutInterval: timeout)
        req.httpMethod = method
        req.setValue(contentType, forHTTPHeaderField: "Content-Type")
        if let token = authToken {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        req.httpBody = body

        let (data, response) = try await URLSession.shared.data(for: req)
        let code = (response as? HTTPURLResponse)?.statusCode ?? 0

        guard code == expectedCode else {
            let detail = (try? JSONSerialization.jsonObject(with: data) as? [String: Any])?["detail"] as? String
                ?? String(data: data, encoding: .utf8)
                ?? "Unknown error"
            throw NSError(domain: "ApiClient", code: code, userInfo: [NSLocalizedDescriptionKey: detail])
        }
        return data
    }

    // MARK: Parsers

    private func parseAuthResponse(_ data: Data) throws -> (String, UserProfile) {
        let json = try JSONSerialization.jsonObject(with: data) as! [String: Any]
        guard let token = json["access_token"] as? String,
              let userJson = json["user"] as? [String: Any] else {
            throw NSError(domain: "ApiClient", code: 0, userInfo: [NSLocalizedDescriptionKey: "Invalid auth response"])
        }
        return (token, try parseUserProfile(userJson))
    }

    private func parseUserProfile(_ json: [String: Any]) throws -> UserProfile {
        UserProfile(
            id: json["id"] as? Int ?? 0,
            email: json["email"] as? String ?? "",
            vehicleType: json["vehicle_type"] as? String ?? "car",
            isDisabled: json["is_disabled"] as? Bool ?? false,
            hasResidentPermit: json["has_resident_permit"] as? Bool ?? false,
            residentZone: json["resident_zone"] as? String ?? ""
        )
    }

    private func parseResult(_ data: Data) throws -> ParkingResult {
        let json = try JSONSerialization.jsonObject(with: data) as! [String: Any]

        let canPark: Bool? = json["can_park"].flatMap {
            if $0 is NSNull { return nil }
            return $0 as? Bool
        }

        func parseStringArray(_ key: String) -> [String] {
            guard let arr = json[key] as? [Any] else { return [] }
            return arr.compactMap { item -> String? in
                if let s = item as? String { return s }
                if let obj = item as? [String: Any] {
                    let text = obj["text"] as? String ?? ""
                    let desc = obj["description"] as? String ?? ""
                    if !text.isEmpty && !desc.isEmpty { return "\(text) - \(desc)" }
                    return desc.isEmpty ? text : desc
                }
                return nil
            }.filter { !$0.isEmpty }
        }

        return ParkingResult(
            canPark: canPark,
            message: json["message"] as? String ?? "",
            notes: parseStringArray("notes"),
            signs: parseStringArray("signs")
        )
    }

    // MARK: Image helpers

    private func resizedJpeg(_ image: UIImage, maxDim: CGFloat = 1024, quality: CGFloat = 0.85) -> Data {
        let size = image.size
        let scale = min(1, maxDim / max(size.width, size.height))
        let newSize = CGSize(width: size.width * scale, height: size.height * scale)

        let renderer = UIGraphicsImageRenderer(size: newSize)
        let resized = renderer.image { _ in image.draw(in: CGRect(origin: .zero, size: newSize)) }
        return resized.jpegData(compressionQuality: quality) ?? Data()
    }
}
